graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/builtins/TuplePrototypeBuiltins.java

```diff
@@ -0,0 +1,995 @@
+                Object value = tuple.getElement(k);
+                boolean selected = toBooleanNode.executeBoolean(
+                        callNode.executeCall(JSArguments.create(thisArg, callbackfn, value, k, tuple))
+                );
+                if (selected) {
+                    list.add(value, growProfile);
+                }
+            }
+            return Tuple.create(list.toArray());
+        }
+    }
+
+    public abstract static class TupleFlattenOperation extends BasicTupleOperation {
+
+        private final BranchProfile errorProfile = BranchProfile.create();
+
+        @Child
+        private JSFunctionCallNode callNode;
+
+        public TupleFlattenOperation(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        protected Object call(Object[] arguments) {
+            if (callNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                callNode = insert(JSFunctionCallNode.createCall());
+            }
+            return callNode.executeCall(arguments);
+        }
+
+        protected void flattenIntoTuple(SimpleArrayList<Object> target, BranchProfile growProfile, Tuple source, long depth, Object mapperFunction, Object thisArg) {
+            for (int i = 0; i < source.getArraySize(); i++) {
+                Object element = source.getElement(i);
+                if (mapperFunction != null) {
+                    element = call(JSArguments.create(thisArg, mapperFunction, element, i, source));
+                    if (JSRuntime.isObject(element)) {
+                        errorProfile.enter();
+                        throw Errors.createTypeError("Tuples cannot contain non-primitive values");
+                    }
+                }
+                if (depth > 0 && element instanceof Tuple) {
+                    flattenIntoTuple(target, growProfile, (Tuple) element, depth - 1, mapperFunction, thisArg);
+                } else {
+                    target.add(element, growProfile);
+                }
+            }
+        }
+    }
+
+    public abstract static class JSTupleFlatNode extends TupleFlattenOperation {
+
+        private final BranchProfile growProfile = BranchProfile.create();
+
+        @Child private JSToIntegerAsLongNode toIntegerNode = JSToIntegerAsLongNode.create();
+
+        public JSTupleFlatNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization
+        protected Tuple flat(Object thisObj, Object depth) {
+            Tuple tuple = thisTupleValue(thisObj);
+            long depthNum = 1;
+            if (depth != Undefined.instance) {
+                depthNum = toIntegerNode.executeLong(depth);
+            }
+            SimpleArrayList<Object> list = SimpleArrayList.create(tuple.getArraySize());
+            flattenIntoTuple(list, growProfile, tuple, depthNum, null, null);
+            return Tuple.create(list.toArray());
```

Comment:
Das _könnte_ ein problem sein. Ich würde mir eine Helper-Methode schreiben, die das `toArray` implementiert, und mit `@TruffleBoundary` markiert ist. außer SimpleArrayList macht das ohnehin (kenne die Klasse nicht).

Reply:
SimpleArrayList is an exists class (not written by me) and it's descriptions states:
> A simple array-based quasi list. Prepared for use-cases in Graal/Truffle to avoid TruffleBoundaries.

As `toArray` results in a singe call to `Arrays.copyOf`, I guess this is fine for PE?
I also checked other `toArray` usages and some aren't `@TruffleBoundary` annotated too...

Reply:
ok, leave it as-is for now. We can fix that later if it is a problem in practise.